File: packages/py/glin_profanity/scanners/composite.py
"""
Composite scanner — runs multiple scanners over a single input in one call.

Mirrors packages/js/src/scanners/composite.ts.

@module scanners/composite
"""

from typing import Optional

from .base import ScanDecision, ScanResult
from .pii import PiiScanner
from .prompt_injection import PromptInjectionScanner
from .secrets import SecretsScanner
from .vault import Vault

# Decision ordering for worst-of aggregation
_DECISION_RANK: dict[ScanDecision, int] = {
    ScanDecision.ALLOW: 0,
    ScanDecision.HITL: 1,
    ScanDecision.BLOCK: 2,
}

_ALL_SCANNERS = ("prompt_injection", "secrets", "pii")


def scan_all(
    text: str,
    scanners: Optional[list[str]] = None,
    vault: Optional[Vault] = None,
) -> list[ScanResult]:
    """Run multiple scanners on text, return results from each.

    Args:
        text: The text to scan.
        scanners: Subset of ``{"prompt_injection", "secrets", "pii"}``.
            Defaults to all three when ``None`` or empty.
        vault: Optional shared Vault instance.  When provided, secrets and
            PII scanners will redact into it (enabling later restoration).

    Returns:
        A list of :class:`ScanResult` objects — one per requested scanner —
        in the order: prompt_injection, secrets, pii.

    Raises:
        ValueError: If an unknown scanner name is supplied.
    """
    active = list(scanners) if scanners else list(_ALL_SCANNERS)

    unknown = set(active) - set(_ALL_SCANNERS)
    if unknown:
        raise ValueError(
            f"Unknown scanner(s): {sorted(unknown)}.  "
            f"Valid names: {list(_ALL_SCANNERS)}"
        )

    redact = vault is not None
    results: list[ScanResult] = []

    for name in _ALL_SCANNERS:
        if name not in active:
            continue

        if name == "prompt_injection":
            scanner = PromptInjectionScanner()
            results.append(scanner.scan(text))

        elif name == "secrets":
            scanner_s = SecretsScanner(redact=redact, vault=vault)
            results.append(scanner_s.scan(text))

        elif name == "pii":
            scanner_p = PiiScanner(redact=redact, vault=vault)
            results.append(scanner_p.scan(text))

    return results


def worst_decision(results: list[ScanResult]) -> ScanDecision:
    """
    Return the most severe :class:`ScanDecision` across a list of results.

    Order: BLOCK > HITL > ALLOW.

    Args:
        results: List of scan results (may be empty).

    Returns:
        The worst decision found, or ``ScanDecision.ALLOW`` for an empty list.
    """
    if not results:
        return ScanDecision.ALLOW
    return max(results, key=lambda r: _DECISION_RANK[r.decision]).decision


__all__ = ["scan_all", "worst_decision"]

File: packages/js/src/scanners/composite.ts
/**
 * Composite scanner — runs multiple scanners over a single input in one call.
 *
 * Mirrors packages/py/glin_profanity/scanners/composite.py.
 *
 * @module scanners/composite
 */

import type { ScanDecision, ScanResult } from './base';
import { PromptInjectionScanner } from './prompt-injection';
import { SecretsScanner } from './secrets';
import { PiiScanner } from './pii';
import type { Vault } from './vault';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** Scanner names supported by scanAll. */
export type CompositeScanner = 'prompt_injection' | 'secrets' | 'pii';

/** Options for scanAll. */
export interface ScanAllOptions {
  /**
   * Subset of scanners to run. Defaults to all three when omitted.
   * Order in the returned array follows the canonical order:
   * prompt_injection → secrets → pii.
   */
  scanners?: CompositeScanner[];
  /**
   * Optional shared Vault instance. When provided, secrets and PII scanners
   * will redact matched values into it (enabling later restoration).
   */
  vault?: Vault;
}

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const ALL_SCANNERS: CompositeScanner[] = ['prompt_injection', 'secrets', 'pii'];

const DECISION_RANK: Record<ScanDecision, number> = {
  ALLOW: 0,
  HITL: 1,
  BLOCK: 2,
};

// ---------------------------------------------------------------------------
// scanAll
// ---------------------------------------------------------------------------

/**
 * Run multiple scanners on text, returning one {@link ScanResult} per scanner.
 *
 * @param text - The text to scan.
 * @param options - Optional configuration (scanner subset + shared vault).
 * @returns Array of ScanResult objects in canonical order:
 *          prompt_injection, secrets, pii.
 */
export function scanAll(text: string, options?: ScanAllOptions): ScanResult[] {
  const requested = options?.scanners && options.scanners.length > 0
    ? options.scanners
    : ALL_SCANNERS;

  const unknown = requested.filter((s) => !(ALL_SCANNERS as string[]).includes(s));
  if (unknown.length > 0) {
    throw new Error(
      `Unknown scanner(s): ${unknown.join(', ')}. Valid names: ${ALL_SCANNERS.join(', ')}`,
    );
  }

  const active = new Set(requested);
  const vault = options?.vault;
  const redact = vault !== undefined;
  const results: ScanResult[] = [];

  for (const name of ALL_SCANNERS) {
    if (!active.has(name)) continue;

    if (name === 'prompt_injection') {
      const scanner = new PromptInjectionScanner();
      results.push(scanner.scan(text));
    } else if (name === 'secrets') {
      const scanner = new SecretsScanner({ redact, vault });
      results.push(scanner.scan(text));
    } else if (name === 'pii') {
      const scanner = new PiiScanner({ redact, vault });
      results.push(scanner.scan(text));
    }
  }

  return results;
}

// ---------------------------------------------------------------------------
// worstDecision
// ---------------------------------------------------------------------------

/**
 * Return the most severe {@link ScanDecision} across a list of results.
 *
 * Order: BLOCK > HITL > ALLOW.
 *
 * @param results - List of scan results (may be empty).
 * @returns The worst decision found, or `'ALLOW'` for an empty list.
 */
export function worstDecision(results: ScanResult[]): ScanDecision {
  if (results.length === 0) return 'ALLOW';
  return results.reduce<ScanDecision>(
    (worst, r) => (DECISION_RANK[r.decision] > DECISION_RANK[worst] ? r.decision : worst),
    'ALLOW',
  );
}
